### PR TITLE
[SPARK-11293] Fix shuffle memory leaks in Spillable collections and UnsafeShuffleWriter (branch-1.5)

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch fixes multiple memory leaks in `Spillable` collections, as well as a leak in `UnsafeShuffleWriter`. There were a small handful of places where tasks would acquire memory from the `ShuffleMemoryManager` but would not release it by the time the task had ended. The `UnsafeShuffleWriter` case was harmless, since the leak could only occur at the very end of a task, but the other two cases are somewhat serious:
- `ExternalSorter.stop()` did not release the sorter's memory. In addition, `BlockStoreShuffleReader` never called `stop()` once the sorter's iterator was fully-consumed. Put together, these bugs meant that a shuffle which performed a reduce-side could starve downstream piplelined transformations of shuffle memory.
- `ExternalAppendOnlyMap` exposes no equivalent of `stop()` and its iterators do not automatically free its in-memory data upon completion. This could cause aggregation operations to starve other operations of shuffle memory.

This patch adds a regression test and fixes all three leaks.

This patch was originally opened as #9260; this version is the 1.5.x backport.

### CODE REVIEW

**Comment by JoshRosen** [comment]
Hmm, looks like legitimate test failures. I'll investigate tomorrow.

File: core/src/main/scala/org/apache/spark/scheduler/Task.scala [hunk 1]
```diff
@@ -89,9 +89,15 @@ private[spark] abstract class Task[T](
     } finally {
       context.markTaskCompleted()
       try {
+        val shuffleMemoryManager = SparkEnv.get.shuffleMemoryManager
+        val shuffleMemoryUsed = shuffleMemoryManager.getMemoryConsumptionForThisTask()
+        if (SparkEnv.get.conf.contains("spark.testing") && shuffleMemoryUsed != 0) {
+          throw new SparkException(
```
Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: In case of failed task, it's fine to have some memory leak (will be freed finally), will this exception override the real one?

**Comment by davies** [comment]
LGTM, except one minor comment.

File: core/src/main/scala/org/apache/spark/scheduler/Task.scala (hunk 1 above)
Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: Ping

**Comment by JoshRosen** [comment]
@davies, I've updated this to address your comment; PTAL.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by davies** [comment]
@JoshRosen  Is there something wrong with this PR?

**Comment by JoshRosen** [comment]
Looks like it timed out while compiling? Let me try again.

Jenkins, retest this please.

**Comment by yhuai** [comment]
tes this please

**Comment by yhuai** [comment]
test this please

**Comment by andrewor14** [comment]
retest this please

**Comment by davies** [comment]
@JoshRosen It should be something wrong in this PR, if we can't fix it easily, would you mind close this one?

**Comment by JoshRosen** [comment]
Yeah, I'm going to close this for now; I don't think that this is a high-priority issue to fix for 1.5.x since it's been around forever and nobody reported problems due to it.
